Inbox unit test fails intermittently

--- FAIL: TestInbox_Error (3.39s)
    --- FAIL: TestInbox_Error/Store_error (0.76s)
        inbox_test.go:256: 
            	Error Trace:	inbox_test.go:256
            	Error:      	Not equal: 
            	            	expected: 200
            	            	actual  : 500
            	Test:       	TestInbox_Error/Store_error

--- a/pkg/activitypub/service/inbox/httpsubscriber/httpsubscriber.go
+++ b/pkg/activitypub/service/inbox/httpsubscriber/httpsubscriber.go
@@ -153,7 +153,7 @@ func (s *Subscriber) respond(msg *message.Message, w http.ResponseWriter, r *htt
 		w.WriteHeader(http.StatusInternalServerError)
 
 	case <-s.stopped:
-		logger.Infof("[%s] Message [%s] wasn not handled since service was stopped", s.ServiceEndpoint, msg.UUID)
+		logger.Infof("[%s] Message [%s] was not handled since service was stopped", s.ServiceEndpoint, msg.UUID)
 
 		s.done <- struct{}{}
 

--- a/pkg/activitypub/service/inbox/inbox.go
+++ b/pkg/activitypub/service/inbox/inbox.go
@@ -159,7 +159,7 @@ func (h *Inbox) handle(msg *message.Message) {
 
 	err := h.jsonUnmarshal(msg.Payload, activity)
 	if err != nil {
-		logger.Errorf("[%s] Error unmarshalling activity message: %s", h.ServiceEndpoint, err)
+		logger.Errorf("[%s] Error unmarshalling activity message [%s]: %s", h.ServiceEndpoint, msg.UUID, err)
 
 		msg.Nack()
 
@@ -169,14 +169,15 @@ func (h *Inbox) handle(msg *message.Message) {
 	activityID, err := h.activityStore.GetActivity(activity.ID().URL())
 	if err != nil {
 		if err != store.ErrNotFound {
-			logger.Errorf("[%s] Error retrieving activity [%s]: %s", h.ServiceEndpoint, activity.ID(), err)
+			logger.Errorf("[%s] Error retrieving activity [%s] in message [%s]: %s",
+				h.ServiceEndpoint, activity.ID(), msg.UUID, err)
 
 			msg.Nack()
 
 			return
 		}
 	} else {
-		logger.Infof("[%s] Ignoring duplicate activity [%s]", h.ServiceEndpoint, activityID)
+		logger.Infof("[%s] Ignoring duplicate activity [%s] in message [%s]", h.ServiceEndpoint, activityID, msg.UUID)
 
 		msg.Nack()
 

--- a/pkg/activitypub/service/inbox/inbox_test.go
+++ b/pkg/activitypub/service/inbox/inbox_test.go
@@ -13,6 +13,7 @@ import (
 	"fmt"
 	"net/http"
 	"net/url"
+	"sync"
 	"testing"
 	"time"
 
@@ -21,6 +22,7 @@ import (
 	"github.com/ThreeDotsLabs/watermill/message"
 	"github.com/google/uuid"
 	"github.com/stretchr/testify/require"
+	"github.com/trustbloc/edge-core/pkg/log"
 	"github.com/trustbloc/sidetree-core-go/pkg/restapi/common"
 
 	"github.com/trustbloc/orb/pkg/activitypub/resthandler"
@@ -145,6 +147,8 @@ func TestInbox_Handle(t *testing.T) {
 }
 
 func TestInbox_Error(t *testing.T) {
+	log.SetLevel("activitypub_service", log.DEBUG)
+
 	client := http.Client{}
 
 	t.Run("Handler error", func(t *testing.T) {
@@ -277,7 +281,23 @@ func TestInbox_Error(t *testing.T) {
 		activityStore := &mocks.ActivityStore{}
 		activityStore.GetActivityReturns(nil, store.ErrNotFound)
 
-		ib, err := New(cfg, activityStore, mocks.NewPubSub(), activityHandler)
+		pubSub := mocks.NewPubSub()
+		undeliverableChan, err := pubSub.Subscribe(context.Background(), spi.UndeliverableTopic)
+		require.NoError(t, err)
+
+		var mutex sync.Mutex
+
+		var undeliverableMessages []*message.Message
+
+		go func() {
+			for msg := range undeliverableChan {
+				mutex.Lock()
+				undeliverableMessages = append(undeliverableMessages, msg)
+				mutex.Unlock()
+			}
+		}()
+
+		ib, err := New(cfg, activityStore, pubSub, activityHandler)
 		require.NoError(t, err)
 		require.NotNil(t, ib)
 
@@ -294,8 +314,6 @@ func TestInbox_Error(t *testing.T) {
 		defer stop()
 
 		time.Sleep(500 * time.Millisecond)
-
-		activityHandler.HandleActivityReturns(fmt.Errorf("injected handler error"))
 
 		activity := vocab.NewCreateActivity(
 			vocab.NewObjectProperty(
@@ -316,6 +334,10 @@ func TestInbox_Error(t *testing.T) {
 		require.NotNil(t, resp)
 		require.Equal(t, http.StatusOK, resp.StatusCode)
 		require.NoError(t, resp.Body.Close())
+
+		mutex.Lock()
+		require.Len(t, undeliverableMessages, 1)
+		mutex.Unlock()
 	})
 
 	t.Run("PubSub subscribe error", func(t *testing.T) {

--- a/pkg/activitypub/service/mocks/mockpubsub.go
+++ b/pkg/activitypub/service/mocks/mockpubsub.go
@@ -76,6 +76,9 @@ func (m *MockPubSub) Publish(topic string, messages ...*message.Message) error {
 	}
 
 	for _, msg := range messages {
+		// Copy the message so that the Ack/Nack is specific to a subscriber
+		msg = msg.Copy()
+
 		msgChan <- msg
 
 		go m.check(msg)

--- a/test/bdd/common_steps.go
+++ b/test/bdd/common_steps.go
@@ -35,7 +35,8 @@ func NewCommonSteps(context *BDDContext, state *state) *CommonSteps {
 }
 
 func (d *CommonSteps) wait(seconds int) error {
-	logger.Infof("Waiting [%d] seconds\n", seconds)
+	logger.Infof("Waiting [%d] seconds", seconds)
+
 	time.Sleep(time.Duration(seconds) * time.Second)
 
 	return nil

